File: app/models/schema_function.rb
class SchemaFunction
  attr_reader :schema_name, :function_name, :language, :return_type, :arg_names, :arg_types, :definition, :description

  def initialize(schema_name, func_name, lang, return_type, arg_names, arg_type, definition, description)
    @schema_name = schema_name
    @function_name = func_name
    @language = lang
    @return_type = return_type
    @arg_names = arg_names
    @arg_types = arg_type
    @definition = definition
    @description = description
  end

  def entity_type_name
    'schema_function'
  end
end

File: app/assets/javascripts/models/schema_function.js
chorus.models.SchemaFunction = chorus.models.Base.extend({
    constructorName: "SchemaFunction",

    toText: function() {
        var functionArguments = this.getFunctionArguments();

        var schemaName = this.ensureDoubleQuoted(this.get("schemaName"));
        var functionName = this.ensureDoubleQuoted(this.get("name"));

        var result = schemaName + "." + functionName + '(';
        result = result + functionArguments.join(', ');
        result = result + ')';
        return result;
    },

    toHintText: function() {
        return this.get("returnType") + " " + this.get('name') + this.formattedArgumentList(true);
    },

    getFunctionArguments: function() {
        var argNames = this.get('argNames') || [];
        return _.map(this.get('argTypes'), function(argType, index) {
            var argName = argNames[index] || "arg" + (index + 1);
            return argType + ' ' + argName;
        });
    },

    formattedArgumentList: function(ensureParams) {
        var args = this.getFunctionArguments();
        if (ensureParams || args.length > 0) {
            return "(" + args.join(", ") + ")";
        } else {
            return "";
        }
    }
});
